src/compiler/checker.ts

```diff
@@ -10258,6 +10260,38 @@ namespace ts {
             return type[cache] = type;
         }
 
+        function getSimplifiedConditionalType(type: ConditionalType, writing: boolean) {
+            const checkType = type.checkType;
+            const extendsType = type.extendsType;
+            const trueType = getTrueTypeFromConditionalType(type);
+            const falseType = getFalseTypeFromConditionalType(type);
+            // Simplifications for types of the form `T extends U ? T : never` and `T extends U ? never : T`.
+            if (falseType.flags & TypeFlags.Never && getActualTypeVariable(trueType) === getActualTypeVariable(checkType)) {
+                if (checkType.flags & TypeFlags.Any || isTypeAssignableTo(getRestrictiveInstantiation(checkType), getRestrictiveInstantiation(extendsType))) { // Always true
+                    return getSimplifiedType(trueType, writing);
+                }
+                else if (isIntersectionEmpty(checkType, extendsType)) { // Always false
+                    return neverType;

```

Comment:
Why wasn't this just a `Debug.fail`?

Reply:
Because it can happen and is perfectly fine when it does. Or am I misunderstanding your question?

Reply:
BTW, the comment "Always false" here refers to when the condition of the conditional type is known to always be false. It doesn't imply that the `if` condition is always false.